### PR TITLE
Feature/tagsystem

### PR DESCRIPTION
**Pull request by MrChrissie**
This feature enhances user experience through the addition of tags for projects.

### CODE REVIEW

File: app/app/settings.py [hunk 1]
```diff
@@ -90,7 +90,7 @@
 TEMPLATES = [
     {
         'BACKEND': 'django.template.backends.django.DjangoTemplates',
-        'DIRS': [path.join(BASE_DIR, 'client/dist')],
+        'DIRS': [path.join(BASE_DIR, 'server/templates')],
```
Comment (fjen) [inline app/app/settings.py]: why?

File: app/api/models.py [hunk 2]
```diff
@@ -230,6 +229,19 @@ class Meta:
...
 
 
+class Tag(models.Model):
+    text = models.TextField()
+    created_at = models.DateTimeField(auto_now_add=True)
+    project = models.ForeignKey(Project, related_name='tags', on_delete=models.CASCADE)
+
+    @property
+    def projectname(self):
+        return self.project.name
```
Comment (fjen) [inline app/api/models.py]: shouldnt be needed

File: app/api/tests/test_api.py [hunk 3]
```diff
@@ -831,6 +831,51 @@ def doCleanups(cls):
...
+                            role_name=settings.ROLE_ANNOTATION_APPROVER)
+        cls.url = reverse(viewname='tag_list')
+        cls.tag = mommy.make('Tag', project=cls.main_project, text='Tag 1')
+
+
+    def test_create_tag(self):
+        self.client.login(username=self.main_project_member_name,
+                        password=self.main_project_member_pass)
+        response = self.client.post(self.url, data={'text': 'Tag 2'})
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
```
Comment (fjen) [inline app/api/tests/test_api.py:862]: add a test to see if tags are listed for a project and another one for tag deletion

File: app/api/admin.py [hunk 4]
```diff
@@ -53,6 +53,13 @@ class RoleMappingAdmin(admin.ModelAdmin):
     ordering = ('user',)
     search_fields = ('user__username',)
 
+
+class TagAdmin(admin.ModelAdmin):
+    list_display = ('text', 'created_at', )
```
Comment (fjen) [inline app/api/admin.py]: add project

File: app/api/serializers.py [hunk 5]
```diff
@@ -119,10 +128,14 @@ def get_current_users_role(self, instance):
                 role_abstractor[key] = users_role["role_id__name"] == val
         return role_abstractor
 
+    def get_tags(self, instance):
+        qs = instance.tags.all()
+        return TagSerializer(qs, many=True).data
+
```
Comment (fjen) [inline app/api/serializers.py]: is this needed?

File: app/api/views/tag.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+class TagList(generics.ListCreateAPIView):
+    serializer_class = TagSerializer
+    pagination_class = None
+
+    def perform_create(self, serializer):
+        project = get_object_or_404(Project, pk=self.kwargs['project_id'])
+        serializer.save(project=project)
+
+    def delete(self, request, *args, **kwargs):
+        delete_ids = request.data['ids']
```
Comment (fjen) [inline app/api/views/tag.py]: usually deletion is only one at a time

File: app/api/serializers.py [hunk 7]
```diff
@@ -122,7 +131,7 @@ def get_current_users_role(self, instance):
     class Meta:
         model = Project
         fields = ('id', 'name', 'description', 'guideline', 'users', 'current_users_role', 'project_type',
-                  'updated_at', 'randomize_document_order', 'collaborative_annotation', 'single_class_classification')
+                  'updated_at', 'randomize_document_order', 'collaborative_annotation', 'single_class_classification', 'tags')
         read_only_fields = ('updated_at', 'users', 'current_users_role')
```
Comment (fjen) [inline app/api/serializers.py]: move tags into `read_only_fields`

File: app/api/tests/test_api.py [hunk 8]
```diff
@@ -831,6 +831,53 @@ def doCleanups(cls):
...
+        self.client.login(username=self.super_user_name,
+                        password=self.super_user_pass)
+        response = self.client.post(self.url, data={'text': 'Tag 2'})
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+
+    def test_tag_list(self):
+        self.client.login(username=self.main_project_member_name,
+                        password=self.main_project_member_pass)
+        response = self.client.get(self.project_url, format='json')
+        self.assertTrue(len(response.data[0]['tags']) == 1 ,'Number of tags differs expected amount.')
```
Comment (fjen) [inline app/api/tests/test_api.py:870]: also check for the right tag text

File: app/api/tests/test_api.py [hunk 9]
```diff
@@ -831,6 +831,53 @@ def doCleanups(cls):
...
+
+    def test_delete_tag(self):
+        self.client.login(username=self.super_user_name,
+                        password=self.super_user_pass)
+        response = self.client.delete(self.delete_url, data={'ids': self.tag.id})
+        self.assertEqual(response.status_code, status.HTTP_204_NO_CONTENT)
```
Comment (fjen) [inline app/api/tests/test_api.py:877]: check if it was really deleted by getting the tag list

File: app/api/tests/test_api.py [hunk 10]
```diff
@@ -831,6 +831,53 @@ def doCleanups(cls):
...
+        cls.tag = mommy.make('Tag', project=cls.main_project, text='Tag 1')
+        cls.url = reverse(viewname='tag_list', args=[cls.main_project.id])
+        cls.project_url = reverse(viewname='project_list')
+        cls.delete_url = reverse(viewname='tag_list', args=[cls.main_project.id])
+
+    def test_create_tag(self):
+        self.client.login(username=self.super_user_name,
+                        password=self.super_user_pass)
+        response = self.client.post(self.url, data={'text': 'Tag 2'})
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
```
Comment (fjen) [inline app/api/tests/test_api.py:862]: check in the tag list

File: app/api/views/tag.py [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+    def get_queryset(self):
+        project = get_object_or_404(Project, pk=self.kwargs['project_id'])
+        return project.tags
+
+    def perform_create(self, serializer):
+        project = get_object_or_404(Project, pk=self.kwargs['project_id'])
+        serializer.save(project=project)
+
+    def delete(self, request, *args, **kwargs):
+        delete_id = request.data['ids']
```
Comment (fjen) [inline app/api/views/tag.py]: `ids`? should be singular

File: frontend/components/project/FormUpdate.vue [hunk 12]
```diff
@@ -134,12 +155,15 @@ import { projectNameRules, descriptionRules } from '@/rules/index'
...
     this.project = await this.$services.project.findById(this.projectId)
+    this.getTags()
   },
 
   data() {
     return {
       project: {},
+      tags: {},
       beforeEditCache: {},
+      input: '',
```
Comment (fjen) [inline frontend/components/project/FormUpdate.vue]: choose a clearer name like `tagText` or `tagInput`

File: frontend/domain/models/label/labelRepository.ts [hunk 13]
```diff
@@ -9,5 +9,5 @@ export interface LabelRepository {
 
   bulkDelete(projectId: string, labelIds: number[]): Promise<void>
 
-  uploadFile(projectId: string, payload: FormData): Promise<void> 
+  uploadFile(projectId: string, payload: FormData): Promise<void>
```
Comment (fjen) [inline frontend/domain/models/label/labelRepository.ts]: no touchy

File: frontend/domain/models/project/project.ts [hunk 14]
```diff
@@ -132,7 +137,7 @@ export class ProjectWriteItem {
     public project_type:                ProjectType,
     public randomize_document_order:    boolean,
     public collaborative_annotation:    boolean,
-    public single_class_classification: boolean
+    public single_class_classification: boolean,
```
Comment (fjen) [inline frontend/domain/models/project/project.ts]: no touchy

File: frontend/domain/models/tag/tag.ts [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+    this.tagItems = this.tagItems.filter(tag => tag.id !== item.id)
+  }
+
+  ids(): Number[]{
+    return this.tagItems.map(item => item.id)
+  }
+
+  get nameList(): string[] {
+    return this.tagItems.map(item => item.name)
+  }
```
Comment (fjen) [inline frontend/domain/models/tag/tag.ts]: not used anywhere

File: frontend/domain/models/tag/tag.ts [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+  static valueOf(
+    { id, text, project }:
+    { id: number, text: string, project: string }
+  ): TagItem {
+    return new TagItem(id, text, project)
+  }
+
+  get name(): string {
+    return this.text
+  }
```
Comment (fjen) [inline frontend/domain/models/tag/tag.ts]: is this used?

File: frontend/repositories/tag/apiTagRepository.ts [hunk 17]
```diff
@@ -0,0 +1,34 @@
+import ApiService from '@/services/api.service'
+import { TagRepository } from '~/domain/models/tag/tagRepository'
+import { TagItem } from '~/domain/models/tag/tag'
+
+export interface TagItemResponse {
+  id: number,
+  text: string,
+  project: string
```
Comment (fjen) [inline frontend/repositories/tag/apiTagRepository.ts:8]: should be a number?

File: frontend/domain/models/tag/tag.ts [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+  toArray(): Object[] {
+    return this.tagItems.map(item => item.toObject())
+  }
+}
+
+export class TagItem {
+  constructor(
+    public id: number,
+    public text: string,
+    public project: string
```
Comment (fjen) [inline frontend/domain/models/tag/tag.ts:29]: `project: number`?

**Review by fjen — CHANGES_REQUESTED** [review]
looks good. only minor things found.

**Comment by fjen** [comment]
please merge the model migrations at the end

File: frontend/repositories/tag/apiTagRepository.ts (hunk 17 above)
Comment (MrChrissie) [reply frontend/repositories/tag/apiTagRepository.ts:8]: project = projectId 
projectId is defined as string in comment methods f.e. - I wanted to keep that scheme.

Comment (fjen) [reply frontend/repositories/tag/apiTagRepository.ts:8]: no problem. i was just wondering

**Pull request closed without merging**